### PR TITLE
Implement Spek JVM runtime

### PR DESCRIPTION
**Pull request by raniejade**
Implementation is pure kotlin and it does not depend on JUnit 5. This is still a work in progress, just want you guys to do the initial review.  Up next is implementing our own gradle/maven plugin.

All previous tests have been ported to the new runtime and all are passing.

### CODE REVIEW

**Comment by raniejade** [comment]
Don't merge this please, this will be the base for next major version of Spek.

**Comment by raniejade** [comment]
@artem-zinnatullin @hhariri @ming13 can you checkout 51ad7c8? Most of spek's engine logic is extracted to **SpekJvmRuntime** and **SpekRuntime**, junit 5 here is acting like a wrapper.

File: build.gradle [hunk 1]
```diff
@@ -15,7 +15,7 @@ buildscript {
         classpath 'org.jetbrains.dokka:dokka-gradle-plugin:0.9.13'
         classpath 'com.jfrog.bintray.gradle:gradle-bintray-plugin:1.6'
         classpath 'org.jfrog.buildinfo:build-info-extractor-gradle:4.1.0'
-        classpath 'org.junit.platform:junit-platform-gradle-plugin:1.0.0-M4'
+        classpath 'org.junit.platform:junit-platform-gradle-plugin:1.0.0-M6'
```
Comment (ming13) [inline build.gradle]: This is kind of breaking, since IJ supports only M5 (I guess?) in the most recent EAP. There is [a version matrix](http://junit.org/junit5/docs/current/user-guide/#running-tests-ide-intellij-idea), but not sure how it is keeping relevant. M5 actually marked failed tests and successful, so I’m hesitant.

File: spek-api/src/main/kotlin/org/jetbrains/spek/api/IgnoreSpek.kt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package org.jetbrains.spek.api
+
+@Retention(AnnotationRetention.RUNTIME)
+@Target(AnnotationTarget.CLASS)
+annotation class IgnoreSpek
```
Comment (ming13) [inline spek-api/src/main/kotlin/org/jetbrains/spek/api/IgnoreSpek.kt]: What is the purpose of this?

File: spek-gradle-plugin/build.gradle [hunk 3]
```diff
@@ -0,0 +1,48 @@
+apply from: "$rootDir/gradle/common/dependencies.gradle"
```
Comment (ming13) [inline spek-gradle-plugin/build.gradle]: Still not 100% onboard on this. We use Spek and (a little bit) JUnit as well. If I understand correctly the plugin adds `spek` task which will ignore JUnit. BTW, shouldn’t this be a separate PR?

File: spek-junit-platform-engine2/src/main/kotlin/org/spekframework/junit/SpekEngineDescriptor.kt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package org.spekframework.junit
```
Comment (ming13) [inline spek-junit-platform-engine2/src/main/kotlin/org/spekframework/junit/SpekEngineDescriptor.kt]: This is a major change, i. e. changing the package name. Shouldn’t it be only in a major version update?

File: build.gradle (hunk 1 above)
Comment (ming13) [inline build.gradle]: Also, submodules use M4 if I’m not mistaken.

File: spek-gradle-plugin/build.gradle (hunk 3 above)
Comment (raniejade) [inline spek-gradle-plugin/build.gradle]: Yep, was just testing out some stuff.

File: spek-api/src/main/kotlin/org/jetbrains/spek/api/IgnoreSpek.kt (hunk 2 above)
Comment (raniejade) [inline spek-api/src/main/kotlin/org/jetbrains/spek/api/IgnoreSpek.kt]: New runtime discovers anonymous non-abstract inner classes, I want to ignore them. Also a fix for #200.

File: spek-junit-platform-engine2/src/main/kotlin/org/spekframework/junit/SpekEngineDescriptor.kt (hunk 4 above)
Comment (raniejade) [inline spek-junit-platform-engine2/src/main/kotlin/org/spekframework/junit/SpekEngineDescriptor.kt]: It's a new module and  package renaming is part of the next major release.

File: build.gradle (hunk 1 above)
Comment (raniejade) [inline build.gradle]: This is just used for running the tests in the repo, I can't remember why I bumped this. Let me take a look.

File: spek-junit-platform-engine2/build.gradle [hunk 5]
```diff
@@ -0,0 +1,35 @@
+apply from: "$rootDir/gradle/common/dependencies.gradle"
+apply from: "$rootDir/gradle/common/kotlin.gradle"
+apply plugin: 'org.junit.platform.gradle.plugin'
+apply plugin: 'org.jetbrains.dokka'
+
+junitPlatform {
+    platformVersion '1.0.0-M4'
```
Comment (artem-zinnatullin) [inline spek-junit-platform-engine2/build.gradle]: Let's put it to `dependencies.gradle` and use same version everywhere?

File: spek-junit-platform-engine2/src/main/kotlin/org/spekframework/junit/SpekEngineDescriptor.kt (hunk 4 above)
Comment (artem-zinnatullin) [inline spek-junit-platform-engine2/src/main/kotlin/org/spekframework/junit/SpekEngineDescriptor.kt]: Then we need to merge this PR to 2.x branch

File: spek-junit-platform-engine2/src/main/kotlin/org/spekframework/junit/RuntimeExecutionListenerAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    override fun executionStart() { }
+
+    override fun executionFinish() { }
+
+    override fun testExecutionStart(test: TestScopeImpl) {
+        listener.executionStarted(test.asDescriptor())
+    }
+
+    override fun testExecutionFinish(test: TestScopeImpl, result: ExecutionResult) {
+        listener.executionFinished(test.asDescriptor(), result.toJunitResult())
```
Comment (artem-zinnatullin) [inline spek-junit-platform-engine2/src/main/kotlin/org/spekframework/junit/RuntimeExecutionListenerAdapter.kt]: Can we allow listener to fail test in this callback? That would be very handy for mimicing JUnit Rules

File: spek-junit-platform-engine2/src/main/kotlin/org/spekframework/junit/TestDescriptorAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    override fun getSource(): Optional<TestSource> {
+        return Optional.empty()
+    }
+
+    override fun removeFromHierarchy() {
+        throw UnsupportedOperationException()
+    }
+
+    override fun setParent(parent: TestDescriptor) {
+        // ignore
```
Comment (artem-zinnatullin) [inline spek-junit-platform-engine2/src/main/kotlin/org/spekframework/junit/TestDescriptorAdapter.kt]: Why? :) More context in the comment will ease support of this code

File: spek-junit-platform-engine2/src/main/kotlin/org/spekframework/junit/TestDescriptorAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    override fun findByUniqueId(uniqueId: UniqueId): Optional<out TestDescriptor> {
+        throw UnsupportedOperationException()
+    }
+
+    override fun getTags(): MutableSet<TestTag> {
+        return mutableSetOf()
+    }
+
+
+    companion object {
```
Comment (artem-zinnatullin) [inline spek-junit-platform-engine2/src/main/kotlin/org/spekframework/junit/TestDescriptorAdapter.kt]: Looks dangerous, let's make it a class and hold reference to it in the engine please

Static access may lead to unexpected usages from everywhere which will make testing and refactoring harder

**Review by artem-zinnatullin — CHANGES_REQUESTED** [review]
Partially reviewed, PR is too big for today, sorry, will continue tomorrow

File: spek-api/src/main/kotlin/org/jetbrains/spek/api/IgnoreSpek.kt (hunk 2 above)
Comment (raniejade) [inline spek-api/src/main/kotlin/org/jetbrains/spek/api/IgnoreSpek.kt]: `Ignore` would be too common and might conflict with other annotations. But probably we could use that since kotlin allows import aliasing.

File: spek-junit-platform-engine2/src/main/kotlin/org/spekframework/junit/RuntimeExecutionListenerAdapter.kt (hunk 6 above)
Comment (raniejade) [inline spek-junit-platform-engine2/src/main/kotlin/org/spekframework/junit/RuntimeExecutionListenerAdapter.kt]: Yeah, let's revisit that later on a different PR.

File: spek-junit-platform-engine2/src/main/kotlin/org/spekframework/junit/TestDescriptorAdapter.kt (hunk 8 above)
Comment (raniejade) [inline spek-junit-platform-engine2/src/main/kotlin/org/spekframework/junit/TestDescriptorAdapter.kt]: :+1:

File: spek-api/src/main/kotlin/org/jetbrains/spek/api/IgnoreSpek.kt (hunk 2 above)
Comment (artem-zinnatullin) [inline spek-api/src/main/kotlin/org/jetbrains/spek/api/IgnoreSpek.kt]: Java has this as well :) Not sure if someone will have problems importing multiple `@Ignore` into same class

File: spek-runtime-test/build.gradle [hunk 9]
```diff
@@ -0,0 +1,8 @@
+apply from: "$rootDir/gradle/common/dependencies.gradle"
+apply from: "$rootDir/gradle/common/kotlin.gradle"
+
+dependencies {
+    compile project(':spek-runtime')
+    compile 'org.jetbrains.kotlin:kotlin-stdlib'
```
Comment (artem-zinnatullin) [inline spek-runtime-test/build.gradle]: version??

File: spek-runtime-test/build.gradle [hunk 10]
```diff
@@ -0,0 +1,8 @@
...
+    compile 'org.jetbrains.kotlin:kotlin-reflect'
```
Comment (artem-zinnatullin) [inline spek-runtime-test/build.gradle]: version??

**Review by artem-zinnatullin — CHANGES_REQUESTED** [review]
Alright, I'm ok with this PR if comments will be fixed and it'll be merged into 2.x branch, then we can cleanup remaining issues

File: spek-runtime-test/build.gradle (hunk 10 above)
Comment (raniejade) [inline spek-runtime-test/build.gradle]: that will be covered in `dependencies.gradle`.  Our build is using https://github.com/spring-gradle-plugins/dependency-management-plugin, to manage dependency versions.

**Comment by raniejade** [comment]
I still have a few changes locally that fixes several issues with JUnit Platform.

**Comment by raniejade** [comment]
Latest commits fixes some issue w/ report generation, cleanup of changes that should be in another PR and the requested changes. Currently I've added a module for the new implementation of the `JUnit TestEngine`, shall I remove the old one and replace it with the new one?

File: spek-runtime/src/main/kotlin/org/spekframework/runtime/scope/ScopeId.kt [hunk 11]
```diff
@@ -0,0 +1,3 @@
+package org.spekframework.runtime.scope
+
+data class ScopeId(val type: String, val name: String)
```
Comment (artem-zinnatullin) [inline spek-runtime/src/main/kotlin/org/spekframework/runtime/scope/ScopeId.kt]: Let's change `type` to sealed class instead of String?

**Comment by artem-zinnatullin** [comment]
I've changed base branch to 2.x, now we really can do breaking changes :)

@ming13, @hhariri PTAL

**Comment by artem-zinnatullin** [comment]
>shall I remove the old one and replace it with the new one?

Let's do it separately so we could review it faster :)

File: spek-runtime/src/main/kotlin/org/spekframework/runtime/scope/ScopeId.kt (hunk 11 above)
Comment (raniejade) [inline spek-runtime/src/main/kotlin/org/spekframework/runtime/scope/ScopeId.kt]: I've used an `enum` instead, since most of the time we will be using `class` and `scope`.

**Comment by raniejade** [comment]
Bumping this up. @artem-zinnatullin

**Review by artem-zinnatullin — APPROVED** [review]

File: spek-junit-platform-engine2/src/main/kotlin/org/spekframework/junit/RuntimeExecutionListenerAdapter.kt (hunk 6 above)
Comment (joelhandwell) [reply spek-junit-platform-engine2/src/main/kotlin/org/spekframework/junit/RuntimeExecutionListenerAdapter.kt:30]: Is there already that different PR ?
